### PR TITLE
domain: support inserting ru historical data into mysql.request_unit_by_group periodically

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve?

Issue Number: ref #49318

Problem Summary:

### What changed and how does it work?
Add a new periodical task to insert daily ru consumption data to table `mysql.request_unit_by_group`. Daily ru consumption is inserted with a periodically task every 24h, and rows older than 61 days will be deleted.
This PR also add a new key in meta store to persist the latest total ru consumption data to help calculate the daily delta value.

NOTE: we uses local timezone to measure the time, it should be ok for single region deployment. We will see if this is a problem for cross-region cluster in the future if we have certain user cases.

Example:(the insert interval is set to 5minutes for test convenience).
```
mysql> select * from mysql.request_unit_by_group limit 5;
+----------------------------+----------------------------+----------------+----------+
| start_time                 | end_time                   | resource_group | total_ru |
+----------------------------+----------------------------+----------------+----------+
| 2023-12-27 18:10:00.000000 | 2023-12-27 18:15:00.000000 | default        |     1361 |
| 2023-12-27 18:10:00.000000 | 2023-12-27 18:15:00.000000 | rg1            |    40043 |
| 2023-12-27 18:10:00.000000 | 2023-12-27 18:15:00.000000 | rg2            |  2235387 |
| 2023-12-27 18:15:00.000000 | 2023-12-27 18:20:00.000000 | default        |     1304 |
| 2023-12-27 18:15:00.000000 | 2023-12-27 18:20:00.000000 | rg2            |   810395 |
+----------------------------+----------------------------+----------------+----------+
5 rows in set (0.02 sec)
```

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Add new system table `mysql.request_unit_by_group` for daily ru consumption data.
```

### CODE REVIEW

File: pkg/domain/ru_stats.go [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+	"github.com/pingcap/tidb/pkg/util/logutil"
+	pd "github.com/tikv/pd/client"
+	"go.uber.org/zap"
+)
+
+const (
+	maxRetryCount   int           = 10
+	ruStatsInterval time.Duration = 24 * time.Hour
+	// only keep stats rows for last 2 monthes(61 days at most).
+	ruStatsGCDuration time.Duration = 61 * ruStatsInterval
```
Comment (CabinfeverB) [inline pkg/domain/ru_stats.go:38]: It's better to keep 61 + 1?

File: pkg/domain/ru_stats.go [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+}
+
+func (r *RUStatsWriter) persistLatestRUStats(stats *meta.RUStats) error {
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnOthers)
+	return kv.RunInNewTxn(ctx, r.store, true, func(ctx context.Context, txn kv.Transaction) error {
+		return meta.NewMeta(txn).SetRUStats(stats)
+	})
+}
+
+func (r *RUStatsWriter) islatestDataInserted(lastEndTime time.Time) (bool, error) {
```
Comment (CabinfeverB) [inline pkg/domain/ru_stats.go:198]:
```suggestion
func (r *RUStatsWriter) isLatestDataInserted(lastEndTime time.Time) (bool, error) {
```

File: pkg/domain/ru_stats.go [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+	end := lastEndTime.Format(time.DateTime)
+	start := lastEndTime.Add(-ruStatsInterval).Format(time.DateTime)
+	rows, sqlErr := execRestrictedSQL(r.sessPool, "SELECT 1 from mysql.request_unit_by_group where start_time = %? and end_time = %? limit 1", []interface{}{start, end})
```
Comment (CabinfeverB) [inline pkg/domain/ru_stats.go:201]: I have some questions about the logic of time judgment in the current implementation. If the writing delay, etc., is likely to occur a time difference greater than 1s, so this logic does not seem to work well

Comment (glorv) [reply pkg/domain/ru_stats.go:201]: Not really, you can see the inserted end_time is the `lastEndTime`, which is the expected end time of yesterday （for 24h duration）not matter what time this task is executed.

File: pkg/domain/ru_stats.go [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+	return &RUStatsWriter{
+		Interval:  ruStatsInterval,
+		RMClient:  do.GetPDClient(),
+		InfoCache: do.infoCache,
+		store:     do.store,
+		sessPool:  do.sysSessionPool,
+	}
+}
+
+func (do *Domain) dailyRequestUnitsWriterLoop() {
```
Comment (nolouch) [inline pkg/domain/ru_stats.go:65]:
```suggestion
func (do *Domain) RequestUnitsWriterLoop() {
```
it relies on `ruStatsInterval`

File: pkg/meta/meta.go [hunk 5]
```diff
@@ -1486,3 +1488,49 @@ func (m *Meta) SetSchemaDiff(diff *model.SchemaDiff) error {
...
+}
+
+// SetRUStats persist new ru stats data to meta storage.
+func (m *Meta) SetRUStats(stats *RUStats) error {
+	data, err := json.Marshal(stats)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = m.txn.Set(mRequestUnitStats, data)
```
Comment (nolouch) [inline pkg/meta/meta.go:1534]: I wonder if this kv will be very large with thousands of groups and many mvcc versions by running long time?

Comment (glorv) [reply pkg/meta/meta.go:1534]: The stats data for a single group is smaller than 1kb, so the stats data should be smaller than 1mb for 1k resource groups, this should be acceptable. Outdated mvcc version should be GCed after new kv is written.

**Review by nolouch — APPROVED** [review]
lgtm

**Review by wjhuang2016 — APPROVED** [review]

**Review by CabinfeverB — APPROVED** [review]

**Pull request merged**